tracer/src/Datadog.Trace/Debugger/SpanCodeOrigin/EndpointDetector.cs

```diff
@@ -0,0 +1,282 @@
+// <copyright file="EndpointDetector.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using Datadog.Trace.VendoredMicrosoftCode.System.Collections.Immutable;
+using Datadog.Trace.VendoredMicrosoftCode.System.Reflection.Metadata;
+using Datadog.Trace.VendoredMicrosoftCode.System.Reflection.Metadata.Ecma335;
+
+namespace Datadog.Trace.Debugger.SpanCodeOrigin;
+
+internal static class EndpointDetector
+{
+    private static readonly HashSet<string> ControllerAttributes =
+    [
+        "Microsoft.AspNetCore.Mvc.ApiControllerAttribute",
+        "Microsoft.AspNetCore.Mvc.ControllerAttribute",
+        "Microsoft.AspNetCore.Mvc.RouteAttribute"
+    ];
+
+    private static readonly HashSet<string> ControllerBaseNames =
+    [
+        "Microsoft.AspNetCore.Mvc.Controller",
+        "Microsoft.AspNetCore.Mvc.ControllerBase"
+    ];
+
+    private static readonly HashSet<string> ActionAttributes =
+    [
+        "Microsoft.AspNetCore.Mvc.HttpGetAttribute",
+        "Microsoft.AspNetCore.Mvc.HttpPostAttribute",
+        "Microsoft.AspNetCore.Mvc.HttpPutAttribute",
+        "Microsoft.AspNetCore.Mvc.HttpDeleteAttribute",
+        "Microsoft.AspNetCore.Mvc.HttpPatchAttribute"
+    ];
+
+    private static readonly HashSet<string> SignalRHubBaseNames =
+    [
+        "Microsoft.AspNetCore.SignalR.Hub",
+        "Microsoft.AspNetCore.SignalR.Hub`1"
+    ];
+
+    private static readonly HashSet<string> PageModelBaseNames = ["Microsoft.AspNetCore.Mvc.RazorPages.PageModel"];
+
+    public static ImmutableHashSet<int> GetEndpointMethodTokens(MetadataReader metadataReader)
```

Comment:
I'm really struggling to follow what this code is doing... The APIs give you a `MethodInfo` (for example) - isn't that sufficient?

Also be aware, that likely none of this works with minimal APIs - i.e. the "default" way to right aspnetcore endpoints these days 😕 I wonder if we could take a completely different approach using action filters and/or endpoint filters/middleware 🤔

Reply:
> I'm really struggling to follow what this code is doing... The APIs give you a MethodInfo (for example) - isn't that sufficient?

We need a pdb info for that method, getting it comes with a price, so I try to get all of the potenional endpoint with a one time analyze per assembly.